invenio_records_resources/services/files/service.py

```diff
@@ -49,11 +50,19 @@ def check_permission(self, identity, action_name, **kwargs):
         return super().check_permission(identity, action_name, **kwargs)
 
     def _get_record(self, id_, identity, action, file_key=None):
-        """Get the associated record."""
+        """Get the associated record.
+
+        If a ``file_key`` is specified and the record in question doesn't have a file
+        for that key, a ``FileKeyNotFoundError`` will be raised.
+        """
         # FIXME: Remove "registered_only=False" since it breaks access to an
         # unpublished record.
         record = self.record_cls.pid.resolve(id_, registered_only=False)
         self.require_permission(identity, action, record=record, file_key=file_key)
+
+        if file_key is not None and file_key not in record.files:
+            raise FileKeyNotFoundError(id_, file_key)
```

Comment:
moving the check into this helper function simplifies the service methods a bit, by reducing duplication.

Reply:
question: I am not sure what the permission check will do or access exactly. Would it make sense to ensure that the file exists before checking permissions?